### ISSUE TITLE
Effects not supported in `custom_vjp` error when using GromovWasserstein

### ISSUE THREAD

**Issue by austinv11** [ott-jax/ott#501]
**Describe the bug**
When attempting to use a Quadratic problem with the GromovWasserstein Solver, I receive a NotImplementedError when the first Sinkhorn iteration converges.  

**Screenshots**
```
  File "C:\Users\Austin\PycharmProjects\riot\riot\model.py", line 82, in fit_dataset
    return self.solver(problem)
  File "C:\Users\Austin\anaconda3\envs\riot\lib\site-packages\ott\solvers\quadratic\gromov_wasserstein.py", line 247, in __call__
    out = iterations(self, prob, init, rng2)
  File "C:\Users\Austin\anaconda3\envs\riot\lib\site-packages\ott\solvers\quadratic\gromov_wasserstein.py", line 415, in iterations
    state = fixed_point_loop.fixpoint_iter(
  File "C:\Users\Austin\anaconda3\envs\riot\lib\site-packages\ott\math\fixed_point_loop.py", line 92, in fixpoint_iter
    _, state = jax.lax.while_loop(max_cond_fn, unrolled_body_fn, (0, state))
  File "C:\Users\Austin\anaconda3\envs\riot\lib\site-packages\ott\math\fixed_point_loop.py", line 80, in unrolled_body_fn
    iteration_state, _ = jax.lax.scan(
  File "C:\Users\Austin\anaconda3\envs\riot\lib\site-packages\ott\math\fixed_point_loop.py", line 76, in one_iteration
    state = body_fn(iteration, constants, state, compute_error)
  File "C:\Users\Austin\anaconda3\envs\riot\lib\site-packages\ott\solvers\quadratic\gromov_wasserstein.py", line 396, in body_fn
    out = solver.linear_ot_solver(linear_pb, init=init)
  File "C:\Users\Austin\anaconda3\envs\riot\lib\site-packages\ott\solvers\linear\sinkhorn.py", line 864, in __call__
    return run(ot_prob, self, (init_dual_a, init_dual_b))
  File "C:\Users\Austin\anaconda3\envs\riot\lib\site-packages\ott\solvers\linear\sinkhorn.py", line 1141, in run
    out = iter_fun(ot_prob, solver, init)
jax._src.source_info_util.JaxStackTraceBeforeTransformation: NotImplementedError: Effects not supported in `custom_vjp`: [<jax._src.callback.IOEffect object at 0x00000279852E2260>]
```

**Desktop (please complete the following information):**
 - OS: Windows 11

**Comment by michalk8** [comment]
Hi @austinv11 , what is your `jax` and `ott-jax` version? From the traceback above, seems like the problem is using IO callbacks (which should be optional and if not passed, not directly affect VJPs).

**Comment by austinv11** [comment]
Hi @michalk8 Here are the versions:
`jax==0.4.25`
`ott-jax==0.4.5`

**Comment by marcocuturi** [comment]
Hi @austinv11 , could you also share the code that led to this error? (not only the error trace)

**Comment by austinv11** [comment]
Sure, here are the relevant code (apologies if it is a little fragmented):

```python
sinkhorn_solver = ott.solvers.linear.sinkhorn.Sinkhorn(
    progress_fn=ott.utils.tqdm_progress_fn(
        pbar=tqdm.tqdm(desc="Sinkhorn", leave=True),
    )
)
solver = ott.solvers.quadratic.gromov_wasserstein.GromovWasserstein(
    relative_epsilon=True,
    linear_ot_solver=sinkhorn_solver,
    progress_fn=ott.utils.tqdm_progress_fn(
        pbar=tqdm.tqdm(desc="Gromov-Wasserstein", leave=True),
    )
)

...

data_X = ... # (n x r)
data_Y = ... # (n x r)

geometry_x = PointCloud(x=data_X, cost_fn=ElasticL1(), relative_epsilon=True)
geometry_y = PointCloud(x=data_Y, cost_fn=ElasticL1(), relative_epsilon=True)
geometry_xy = PointCloud(x=data_X, y=data_Y, cost_fn=ElasticL1(), relative_epsilon=True)

a = data_X.sum(axis=1) / data_X.sum()
b = data_Y.sum(axis=1) / data_Y.sum()
tau_a = 0.95  # I have also tried setting this to 1
tau_b = 1.0
problem = QuadraticProblem(
    geom_xx=geometry_x,
    geom_yy=geometry_y,
    geom_xy=geometry_xy,
    scale_cost=True,
    a=a,
    b=b,
    loss="sqeucl",  # kl
    tau_a=tau_a,
    tau_b=tau_b
)

res = solver(problem)
```

**Comment by marcocuturi** [comment]
just to be sure, does removing the tqdm callback change anything?

**Comment by austinv11** [comment]
Hello @marcocuturi , removing the progress bar does indeed fix it. Though I do miss being able to tell what is going on during the run time.

**Comment by marcocuturi** [comment]
in that case, does setting 

```
sinkhorn_solver = ott.solvers.linear.sinkhorn.Sinkhorn(
    implicit_diff=None,
    progress_fn=ott.utils.tqdm_progress_fn(
        pbar=tqdm.tqdm(desc="Sinkhorn", leave=True),
    )
)
```
solve the problem?

**Comment by marcocuturi** [comment]
(note aside, in your definition 
```
data_X = ... # (n x r)
data_Y = ... # (n x r)

geometry_x = PointCloud(x=data_X, cost_fn=ElasticL1(), relative_epsilon=True)
geometry_y = PointCloud(x=data_Y, cost_fn=ElasticL1(), relative_epsilon=True)
geometry_xy = PointCloud(x=data_X, y=data_Y, cost_fn=ElasticL1(), relative_epsilon=True)

a = data_X.sum(axis=1) / data_X.sum()
b = data_Y.sum(axis=1) / data_Y.sum()
```
I am not sure why the weights and the points on which you create your geometries come from the same matrix... but maybe this was just to formulate an example :) ?

**Comment by austinv11** [comment]
Hello @marcocuturi unfortunately neither work. 

And thanks for your question. It was partially to make an example as the data that go into the point clouds are processed differently rather than being directly inputted. Though it is possible that I misunderstood when `a` and `b` should be specified instead of leaving to defaults.

**Comment by marcocuturi** [comment]
and if you also fix the iterations of gromov? to the same min/max?

**Comment by austinv11** [comment]
Hello @marcocuturi , apologies for the delayed response. I have tried fixing both to the same min/max, and I still get the same error unfortunately

**Comment by michalk8** [comment]
@austinv11 should be fixed via #506 on `main`, will try to make release in the upcoming days

**Comment by austinv11** [comment]
Amazing, I appreciate your responsiveness!
